File: [0547][Medium][Friend_Circles]/Friend_Circles_2.py
# solution 1 greedy algorithm.
# leetcode time     cost : 216 ms
# leetcode memory   cost : 14.6 MB 
# Time  Complexity: O(n**2)
# Space Complexity: O(1)
# solution 1, DFS,
class Solution:
    def findCircleNum(self, matrix_M):
        N = len(matrix_M)
        visited = set()
        def dfs(node):
            for j, friend_i_j in enumerate(matrix_M[node]):  #dfs(i), nei is col j, adj is the matrix_M[i][j],
                if friend_i_j and (j not in visited):
                    visited.add(j)
                    dfs(j)
        
        ans = 0
        for i in range(N):
            if i not in visited:
                dfs(i)
                ans += 1
        return ans
    
def main():
    M = [[1,1,0],[1,1,0],[0,0,1]]
    obj = Solution()
    res = obj.findCircleNum(M)
    print("return value is ",res)
    
if __name__ =='__main__':
    main()     

File: [0547][Medium][Friend_Circles]/Friend_Circles_2.java
// leetcode time     cost : 8 ms
// leetcode memory   cost : 41.4 MB 
// Time  Complexity: O(n**2)
// Space Complexity: O(n)
// solution 2, BFS,
import java.util.LinkedList;
import java.util.Queue;
class Solution {
    public int findCircleNum(int[][] M) {
        int[] visited = new int[M.length];
        int count = 0;
        Queue < Integer > queue = new LinkedList < > ();
        for (int i = 0; i < M.length; i++) {
            if (visited[i] == 0) {
                queue.add(i);
                while (!queue.isEmpty()) {
                    int s = queue.remove();
                    // visited[s] = 1; the same effect as the statement in line 21,
                    for (int j = 0; j < M.length; j++) {
                        if (M[s][j] == 1 && visited[j] == 0)
                            visited[s] = 1;
                            queue.add(j);
                    }
                }
                count++;
            }
        }
        return count;
    }
}

public class Friend_Circles_2 {
    public static void main(String args[]) {
        int[][] M = {{1,1,0},{1,1,0},{0,0,1}}; // expect is 2,
        Solution Solution_obj = new Solution();
        int result = Solution_obj.findCircleNum(M);
        System.out.println("In java code,result is :" + result);
    }
}